Parse.FacebookUtils.logIn is creating new users when Facebook ID already exists.

We have been using a very simple flow to log in users using Facebook that has been working for almost a year and a half, and now it has stopped working.

We use Parse.FacebookUtils.logIn to log in an existing user, or creating a new one. 

```
var facebookAuthData = {
    "id": <fb_id>,
    "access_token": <fb_access_token>,
    "expiration_date": <fb_expiration_date>
};
Parse.FacebookUtils.logIn(facebookAuthData, {
    success: function(user) {
        if (user.isNew()) {
            // New user....
        } else {
            // Old user...
        }
    },
    error: ...
});
```

But it's happening that although there is an existing user in the database with the exact <fb_id> that we pass to Parse.FacebookUtils.logIn(), it is creating a new user instead of login the existing user. Furthermore, this "new" user is created with a weird/random "username" in the database.

Some more details:

- Parse server version: 2.8.2. FB Graph version 3.0.

- We only see the issue in our "Production" environment. This is a Heroku dyno with a mongoLab add-on. Note that it uses a replica set.

- Our development environment uses a free-tier Heroku dyno, with a mongoLab add-on without a replica set. We DON'T see the issue here.

- New users that have never logged in, can successfully log-in using Facebook, and re-login Ok. The ones that are having trouble were old users that logged in through Facebook a while back (using FB Graph v2.2).

- After Parse.FacebookUtils.login() returns a "new" user when it shouldn't, when we try to save to the new user (i.e. to set the email), the error returned by parse can be inconsistent. Sometimes it will say "code=203, message=Account already exists for this email address." (since we had previously created the user and assigned the email address), but if I manually delete the malformed new user, sometimes instead of creating another malformed user, it will just say "Invalid access token". Very weird.

- I've tried with older versions of Parse Server too (like 2.7.4) and the same issue happens. 

I've tried for days trying to understand why this is happening with no luck. Any help appreciated, especially since this one reached our Production environment and previously FB logged users are being locked out.

Thanks in advance for the help.

Extract of our package.json:

```
{
  "dependencies": {
    "@parse/simple-mailgun-adapter": "^1.0.2",
    "axios": "^0.18.0",
    "bcryptjs": "^2.4.3",
    "express": "~4.16.x",
    "kerberos": "^0.0.24",
    "lodash": "^4.17.10",
    "moment": "^2.22.2",
    "nan": "^2.10.0",
    "node-gyp": ">=3.7.0",
    "node-pre-gyp": "^0.10.2",
    "npm": ">=6.1.0",
    "parse": "~1.11.1",
    "parse-dashboard": "~1.2.0",
    "parse-server": "~2.8.2",
    "pre-commit": "^1.2.2",
    "request": "^2.87.0",
    "request-promise": "^4.2.2",
    "semver": "^5.5.0",
    "socks": "~2.2.0",
    "underscore": "~1.9.1"
  },
  "scripts": {
    "start": "node index.js",
    ...
  },
  "engines": {
    "node": ">=10.0.0"
  }
}
```

Here is the log when Parse is deployed (kerberos compiles with warnings):

```
$ git push heroku v1.8.2:master -f
Total 0 (delta 0), reused 0 (delta 0)
remote: Compressing source files... done.
remote: Building source:
remote:
remote: -----> Node.js app detected
remote:
remote: -----> Creating runtime environment
remote:
remote:        NPM_CONFIG_LOGLEVEL=error
remote:        NODE_VERBOSE=false
remote:        NODE_ENV=production
remote:        NODE_MODULES_CACHE=false
remote:
remote: -----> Installing binaries
remote:        engines.node (package.json):  >=10.0.0
remote:        engines.npm (package.json):   unspecified (use default)
remote:
remote:        Resolving node version >=10.0.0...
remote:        Downloading and installing node 10.6.0...
remote:        Using default npm version: 6.1.0
remote:
remote: -----> Restoring cache
remote:        Skipping cache restore (disabled)
remote:
remote: -----> Building dependencies
remote:        Installing node modules (package.json + package-lock)
remote:
remote:        > bcrypt@2.0.1 install /tmp/build_bd81344b7effebfd0db2762cdf22abea/node_modules/bcrypt
remote:        > node-pre-gyp install --fallback-to-build
remote:
remote:        [bcrypt] Success: "/tmp/build_bd81344b7effebfd0db2762cdf22abea/node_modules/bcrypt/lib/binding/bcrypt_lib.node" is installed via remote
remote:
remote:        > kerberos@0.0.24 install /tmp/build_bd81344b7effebfd0db2762cdf22abea/node_modules/kerberos
remote:        > (node-gyp rebuild) || (exit 0)
remote:
remote:        make: Entering directory '/tmp/build_bd81344b7effebfd0db2762cdf22abea/node_modules/kerberos/build'
remote:        CXX(target) Release/obj.target/kerberos/lib/kerberos.o
remote:        ../lib/kerberos.cc: In static member function ‘static void Kerberos::After(uv_work_t*)’:
remote:        ../lib/kerberos.cc:948:50: warning: ‘v8::Local<v8::Value> Nan::Callback::Call(int, v8::Local<v8::Value>*) const’ is deprecated [-Wdeprecated-declarations]
remote:        worker->callback->Call(ARRAY_SIZE(info), info);
remote:        ^
remote:        In file included from ../lib/kerberos.h:9:0,
remote:        from ../lib/kerberos.cc:1:
remote:        ../../nan/nan.h:1618:3: note: declared here
remote:        Call(int argc, v8::Local<v8::Value> argv[]) const {
remote:        ^
remote:        ../lib/kerberos.cc:972:50: warning: ‘v8::Local<v8::Value> Nan::Callback::Call(int, v8::Local<v8::Value>*) const’ is deprecated [-Wdeprecated-declarations]
remote:        worker->callback->Call(ARRAY_SIZE(info), info);
remote:        ^
remote:        In file included from ../lib/kerberos.h:9:0,
remote:        from ../lib/kerberos.cc:1:
remote:        ../../nan/nan.h:1618:3: note: declared here
remote:        Call(int argc, v8::Local<v8::Value> argv[]) const {
remote:        ^
remote:        CXX(target) Release/obj.target/kerberos/lib/worker.o
remote:        CC(target) Release/obj.target/kerberos/lib/kerberosgss.o
remote:        ../lib/kerberosgss.c:36:0: warning: ignoring #pragma clang diagnostic [-Wunknown-pragmas]
remote:        #pragma clang diagnostic push
remote:        ^
remote:        ../lib/kerberosgss.c:37:0: warning: ignoring #pragma clang diagnostic [-Wunknown-pragmas]
remote:        #pragma clang diagnostic ignored "-Wdeprecated-declarations"
remote:        ^
remote:        ../lib/kerberosgss.c: In function ‘authenticate_gss_client_wrap’:
remote:        ../lib/kerberosgss.c:446:19: warning: variable ‘server_conf_flags’ set but not used [-Wunused-but-set-variable]
remote:        char buf[4096], server_conf_flags;
remote:        ^
remote:        ../lib/kerberosgss.c: At top level:
remote:        ../lib/kerberosgss.c:1236:0: warning: ignoring #pragma clang diagnostic [-Wunknown-pragmas]
remote:        #pragma clang diagnostic pop
remote:        ^
remote:        CC(target) Release/obj.target/kerberos/lib/base64.o
remote:        CXX(target) Release/obj.target/kerberos/lib/kerberos_context.o
remote:        SOLINK_MODULE(target) Release/obj.target/kerberos.node
remote:        COPY Release/kerberos.node
remote:        make: Leaving directory '/tmp/build_bd81344b7effebfd0db2762cdf22abea/node_modules/kerberos/build'
remote:
remote:        > pre-commit@1.2.2 install /tmp/build_bd81344b7effebfd0db2762cdf22abea/node_modules/pre-commit
remote:        > node install.js
remote:
remote:
remote:        > spawn-sync@1.0.15 postinstall /tmp/build_bd81344b7effebfd0db2762cdf22abea/node_modules/spawn-sync
remote:        > node postinstall
remote:
remote:
remote:        > nodemon@1.17.5 postinstall /tmp/build_bd81344b7effebfd0db2762cdf22abea/node_modules/nodemon
remote:        > node bin/postinstall || exit 0
remote:
remote:        Love nodemon? You can now support the project via the open collective:
remote:        > https://opencollective.com/nodemon/donate
remote:
remote:
remote:        > parse-server@2.8.2 postinstall /tmp/build_bd81344b7effebfd0db2762cdf22abea/node_modules/parse-server
remote:        > node -p 'require("./postinstall.js")()'
remote:
remote:
remote:        1111111111
remote:        1111111111111111
remote:        1111111111111111111111
remote:        11111111111111111111111111
remote:        111111111111111       11111111
remote:        1111111111111             111111
remote:        1111111111111   111111111   111111
remote:        111111111111   11111111111   111111
remote:        1111111111111   11111111111   111111
remote:        1111111111111   1111111111    111111
remote:        1111111111111111111111111    1111111
remote:        11111111                    11111111
remote:        111111         1111111111111111111
remote:        11111   11111  111111111111111111
remote:        11111         11111111111111111
remote:        111111     111111111111111111
remote:        11111111111111111111111111
remote:        1111111111111111111111
remote:        111111111111111111
remote:        11111111111
remote:
remote:
remote:        Thanks for installing parse 🙏
remote:        Please consider donating to our open collective
remote:        to help us maintain this package.
remote:
remote:        👉 https://opencollective.com/parse-server
remote:
remote:        added 1423 packages from 1491 contributors and audited 8063 packages in 36.955s
remote:        found 8 vulnerabilities (4 low, 4 moderate)
remote:        run `npm audit fix` to fix them, or `npm audit` for details
remote:
remote: -----> Caching build
remote:        Clearing previous node cache
remote:        Skipping cache save (disabled by config)
remote:
remote: -----> Pruning devDependencies
remote:        removed 220 packages and audited 6620 packages in 8.974s
remote:        found 7 vulnerabilities (3 low, 4 moderate)
remote:        run `npm audit fix` to fix them, or `npm audit` for details
remote:
remote: -----> Build succeeded!
remote: -----> Discovering process types
remote:        Procfile declares types     -> (none)
remote:        Default types for buildpack -> web
remote:
remote: -----> Compressing...
remote:        Done: 45.1M
remote: -----> Launching...
remote:        Released v143
remote:        <my heroku app> deployed to Heroku
remote:
remote: Verifying deploy... done.
To <my heroku remote repo>
 + 7905cfe...9915b97 v1.8.2 -> master (forced update)
```

When the existing user tries to logging to facebook (a user that already has a FB auth data well defined), the Parse.FacebookUtils.logIn() utility will return a new user with random username (expressed in the log below: Q2NGibf9AGcl3W7s6Ud881TiX), instead of returning the existing one which has the same authData.facebook.id. Note that the "Account already exists for this email address." happens after I have received the undesired new user and assign it an email. This is what it logged during the FB login attempt:

```
2018-07-17T06:29:25.742808+00:00 app[web.1]: [36mverbose[39m: REQUEST for [POST] /users: {
2018-07-17T06:29:25.742822+00:00 app[web.1]:   "authData": {
2018-07-17T06:29:25.742824+00:00 app[web.1]:     "facebook": {
2018-07-17T06:29:25.742826+00:00 app[web.1]:       "id": "<fb_id>",
2018-07-17T06:29:25.742831+00:00 app[web.1]:       "access_token": "<fb_token>",
2018-07-17T06:29:25.742833+00:00 app[web.1]:       "expiration_date": "2026-08-17T17:35:02.659Z"
2018-07-17T06:29:25.742835+00:00 app[web.1]:     }
2018-07-17T06:29:25.742836+00:00 app[web.1]:   }
2018-07-17T06:29:25.742838+00:00 app[web.1]: } method=POST, url=/users, host=<my_heroku_app>, connection=close, content-type=text/plain, origin=file://, if-none-match=W/"3b7-v0XpDL5Ht2PFXiJZ4QnlSoxPaOI", accept=*/*, user-agent=Mozilla/5.0 (iPhone; CPU iPhone OS 11_4 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Mobile/15F79, accept-language=es-es, accept-encoding=br, gzip, deflate, x-request-id=81938098-5cdc-4be0-960e-0b9a2f486dbb, x-forwarded-for=<ip>, x-forwarded-proto=https, x-forwarded-port=443, via=1.1 vegur, connect-time=1, x-request-start=1531808965738, total-route-time=0, content-length=406, id=<fb_id>, access_token=<fb_token>, expiration_date=2026-08-17T17:35:02.659Z
2018-07-17T06:29:25.837790+00:00 app[web.1]: [36mverbose[39m: RESPONSE from [POST] /users: {
2018-07-17T06:29:25.837795+00:00 app[web.1]:   "status": 201,
2018-07-17T06:29:25.837797+00:00 app[web.1]:   "response": {
2018-07-17T06:29:25.837799+00:00 app[web.1]:     "objectId": "9Rrk8Xn8HJ",
2018-07-17T06:29:25.837804+00:00 app[web.1]:     "createdAt": "2018-07-17T06:29:25.745Z",
2018-07-17T06:29:25.837806+00:00 app[web.1]:     "username": "Q2NGibf9AGcl3W7s6Ud881TiX",
2018-07-17T06:29:25.837808+00:00 app[web.1]:     "sessionToken": "r:<some_session_token>"
2018-07-17T06:29:25.837809+00:00 app[web.1]:   },
2018-07-17T06:29:25.837811+00:00 app[web.1]:   "location": "https://<my_heroku_app>/users/9Rrk8Xn8HJ"
2018-07-17T06:29:25.837814+00:00 app[web.1]: } status=201, objectId=9Rrk8Xn8HJ, createdAt=2018-07-17T06:29:25.745Z, username=Q2NGibf9AGcl3W7s6Ud881TiX, sessionToken=r:<some_session_token>, location=https://<my_heroku_app>/users/9Rrk8Xn8HJ
2018-07-17T06:29:25.839443+00:00 heroku[router]: at=info method=POST path="/users" host=<my_heroku_app> request_id=81938098-5cdc-4be0-960e-0b9a2f486dbb fwd="<ip>" dyno=web.1 connect=1ms service=99ms status=201 bytes=751 protocol=https
2018-07-17T06:29:26.023821+00:00 app[web.1]: [36mverbose[39m: REQUEST for [PUT] /classes/_User/9Rrk8Xn8HJ: {
2018-07-17T06:29:26.023839+00:00 app[web.1]:   "nombre": "<some_name>",
2018-07-17T06:29:26.023841+00:00 app[web.1]:   "email": "<some_email>",
2018-07-17T06:29:26.023843+00:00 app[web.1]:   "imageUrl": "<some_img_url>",
2018-07-17T06:29:26.023845+00:00 app[web.1]:   "genero": "F"
2018-07-17T06:29:26.023847+00:00 app[web.1]: } method=PUT, url=/classes/_User/9Rrk8Xn8HJ, host=<my_heroku_app>, connection=close, content-type=text/plain, origin=file://, accept=*/*, user-agent=Mozilla/5.0 (iPhone; CPU iPhone OS 11_4 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Mobile/15F79, accept-language=es-es, accept-encoding=br, gzip, deflate, x-request-id=1301cc6b-158f-4a4c-82e1-0f1343c83ac2, x-forwarded-for=186.15.145.21, x-forwarded-proto=https, x-forwarded-port=443, via=1.1 vegur, connect-time=1, x-request-start=1531808966003, total-route-time=0, content-length=410, nombre=<some_name>, email=<some_email>, imageUrl=<some_img_url>, genero=F
2018-07-17T06:29:26.052244+00:00 app[web.1]: [31merror[39m: Error generating response. ParseError {
2018-07-17T06:29:26.052247+00:00 app[web.1]:   code: 203,
2018-07-17T06:29:26.052252+00:00 app[web.1]:   message: 'Account already exists for this email address.' } code=203, message=Account already exists for this email address.
2018-07-17T06:29:26.053969+00:00 app[web.1]: [31merror[39m: Account already exists for this email address. code=203, message=Account already exists for this email address.
2018-07-17T06:29:26.054682+00:00 heroku[router]: at=info method=POST path="/classes/_User/9Rrk8Xn8HJ" host=<my_heroku_app> request_id=1301cc6b-158f-4a4c-82e1-0f1343c83ac2 fwd="<ip>" dyno=web.1 connect=1ms service=49ms status=400 bytes=602 protocol=https
```

I cannot share my environment, but I can confirm that the issue can be reproduced consistently with Parse v2.8.2, just let me know how I can help. Right now I've deployed 2.7.4 to avoid the issue in Production.

Some notes about the environment:

- This deployment is running MongoDB version 3.0.15 (MMAPv1). This is a standard heroku production deployment which has been working for about year and a half with no intervention other than Parse Server updates.

- rs.conf() raw data output:

```
{
  "_id": "<some_id>",
  "version": 3,
  "members": [
    {
      "_id": 0,
      "host": "<some_host>",
      "arbiterOnly": false,
      "buildIndexes": true,
      "hidden": false,
      "priority": 1,
      "tags": {},
      "slaveDelay": 0,
      "votes": 1
    },
    {
      "_id": 1,
      "host": "<some_host>",
      "arbiterOnly": false,
      "buildIndexes": true,
      "hidden": false,
      "priority": 1,
      "tags": {},
      "slaveDelay": 0,
      "votes": 1
    },
    {
      "_id": 10,
      "host": "<some_host>",
      "arbiterOnly": true,
      "buildIndexes": true,
      "hidden": false,
      "priority": 1,
      "tags": {},
      "slaveDelay": 0,
      "votes": 1
    }
  ],
  "settings": {
    "chainingAllowed": true,
    "heartbeatTimeoutSecs": 10,
    "getLastErrorModes": {},
    "getLastErrorDefaults": {
      "w": 1,
      "wtimeout": 0
    }
  }
}
```

- rs.status() raw output:

```
{
  "set": "rs-<some_id>",
  "date": {
    "$date": "2018-07-17T06:49:04.723Z"
  },
  "myState": 1,
  "members": [
    {
      "_id": 0,
      "name": "<some_name>",
      "health": 1,
      "state": 1,
      "stateStr": "PRIMARY",
      "uptime": 75758,
      "optime": {
        "$ts": 1531809010,
        "$inc": 1
      },
      "optimeDate": {
        "$date": "2018-07-17T06:30:10.000Z"
      },
      "electionTime": {
        "$ts": 1531734394,
        "$inc": 1
      },
      "electionDate": {
        "$date": "2018-07-16T09:46:34.000Z"
      },
      "configVersion": 3,
      "self": true,
      "serverId": "<some_server_id>"
    },
    {
      "_id": 1,
      "name": "<some_name>",
      "health": 1,
      "state": 2,
      "stateStr": "SECONDARY",
      "uptime": 75757,
      "optime": {
        "$ts": 1531809010,
        "$inc": 1
      },
      "optimeDate": {
        "$date": "2018-07-17T06:30:10.000Z"
      },
      "lastHeartbeat": {
        "$date": "2018-07-17T06:49:03.133Z"
      },
      "lastHeartbeatRecv": {
        "$date": "2018-07-17T06:49:03.448Z"
      },
      "pingMs": 0,
      "syncingTo": "<some_name>",
      "configVersion": 3,
      "serverId": "s-ds025502-a1"
    },
    {
      "_id": 10,
      "name": "<some_name>,
      "health": 1,
      "state": 7,
      "stateStr": "ARBITER",
      "uptime": 75757,
      "lastHeartbeat": {
        "$date": "2018-07-17T06:49:04.078Z"
      },
      "lastHeartbeatRecv": {
        "$date": "2018-07-17T06:49:03.854Z"
      },
      "pingMs": 16,
      "configVersion": 3,
      "serverId": "s-ds025502-ar0"
    }
  ],
  "ok": 1
}
```

The _User collection has 1 index:

```
db['system.indexes'].find() Raw Output
{
  "v": 1,
  "key": {
    "_id": 1
  },
  "name": "_id_",
  "ns": "<some_name>._User"
}
```

--- a/spec/ParseUser.spec.js
+++ b/spec/ParseUser.spec.js
@@ -3706,4 +3706,36 @@ describe('Parse.User testing', () => {
       expect(results.length).toBe(1);
     }).then(done, done.fail);
   });
+
+  describe('issue #4897', () => {
+    it("should be able to login with a legacy user (no ACL)", async () => {
+      // This issue is a side effect of the locked users and legacy users which don't have ACL's
+      // In this scenario, a legacy user wasn't be able to login as there's no ACL on it
+      const database = Config.get(Parse.applicationId).database;
+      const collection = await database.adapter._adaptiveCollection('_User');
+      await collection.insertOne({
+        "_id": "ABCDEF1234",
+        "name": "<some_name>",
+        "email": "<some_email>",
+        "username": "<some_username>",
+        "_hashed_password": "<some_password>",
+        "_auth_data_facebook": {
+          "id": "8675309",
+          "access_token": "jenny"
+        },
+        "sessionToken": "<some_session_token>",
+      });
+      const provider = getMockFacebookProvider();
+      Parse.User._registerAuthenticationProvider(provider);
+      const model = await Parse.User._logInWith("facebook", {});
+      expect(model.id).toBe('ABCDEF1234');
+      ok(model instanceof Parse.User, "Model should be a Parse.User");
+      strictEqual(Parse.User.current(), model);
+      ok(model.extended(), "Should have used subclass.");
+      strictEqual(provider.authData.id, provider.synchronizedUserId);
+      strictEqual(provider.authData.access_token, provider.synchronizedAuthToken);
+      strictEqual(provider.authData.expiration_date, provider.synchronizedExpiration);
+      ok(model._isLinked("facebook"), "User should be linked to facebook");
+    });
+  });
 });

--- a/spec/ParseUser.spec.js
+++ b/spec/ParseUser.spec.js
@@ -3708,7 +3708,7 @@ describe('Parse.User testing', () => {
   });
 
   describe('issue #4897', () => {
-    it("should be able to login with a legacy user (no ACL)", async () => {
+    it_only_db('mongo')("should be able to login with a legacy user (no ACL)", async () => {
       // This issue is a side effect of the locked users and legacy users which don't have ACL's
       // In this scenario, a legacy user wasn't be able to login as there's no ACL on it
       const database = Config.get(Parse.applicationId).database;

--- a/src/Controllers/DatabaseController.js
+++ b/src/Controllers/DatabaseController.js
@@ -427,7 +427,7 @@ class DatabaseController {
                   }
                 });
                 for (const updateOperation in update) {
-                  if (Object.keys(updateOperation).some(innerKey => innerKey.includes('$') || innerKey.includes('.'))) {
+                  if (update[updateOperation] && typeof update[updateOperation] === 'object' && Object.keys(update[updateOperation]).some(innerKey => innerKey.includes('$') || innerKey.includes('.'))) {
                     throw new Parse.Error(Parse.Error.INVALID_NESTED_KEY, "Nested keys should not contain the '$' or '.' characters");
                   }
                 }
@@ -660,7 +660,7 @@ class DatabaseController {
    * @param {boolean} fast set to true if it's ok to just delete rows and not indexes
    * @returns {Promise<void>} when the deletions completes
    */
-  deleteEverything(fast: boolean = false): Promise<void> {
+  deleteEverything(fast: boolean = false): Promise<any> {
     this.schemaPromise = null;
     return Promise.all([
       this.adapter.deleteAllClasses(fast),

--- a/src/RestWrite.js
+++ b/src/RestWrite.js
@@ -278,13 +278,23 @@ RestWrite.prototype.findUsersWithAuthData = function(authData) {
   return findPromise;
 }
 
+RestWrite.prototype.filteredObjectsByACL = function(objects) {
+  if (this.auth.isMaster) {
+    return objects;
+  }
+  return objects.filter((object) => {
+    if (!object.ACL) {
+      return true; // legacy users that have no ACL field on them
+    }
+    // Regular users that have been locked out.
+    return object.ACL && Object.keys(object.ACL).length > 0;
+  });
+}
 
 RestWrite.prototype.handleAuthData = function(authData) {
   let results;
   return this.findUsersWithAuthData(authData).then((r) => {
-    results = r.filter((user) => {
-      return !this.auth.isMaster && user.ACL && Object.keys(user.ACL).length > 0;
-    });
+    results = this.filteredObjectsByACL(r);
     if (results.length > 1) {
       // More than 1 user with the passed id's
       throw new Parse.Error(Parse.Error.ACCOUNT_ALREADY_LINKED,

